src/wp-includes/class-wp-block-type.php

```diff
@@ -522,6 +522,20 @@ public function set_props( $args ) {
 		$args = apply_filters( 'register_block_type_args', $args, $this->name );
 
 		foreach ( $args as $property_name => $property_value ) {
+			// Avoid infinite recursion in block hooks (hooking to itself).
+			if ( 'block_hooks' === $property_name ) {
+				if ( ! is_array( $property_value ) ) {
+					continue;
+				}
+				if ( array_key_exists( $this->name, $property_value ) ) {
+					_doing_it_wrong(
+						__METHOD__,
+						__( 'Cannot hook block to itself.' ),
+						'6.4.0'
+					);
+					unset( $property_value[ $this->name ] );
+				}
+			}
```

Comment:
Sorry, I really think we should keep this in `register_block_type_from_metadata` 😬 AFAICS, no other field is being validated here in `WP_Block_Type` -- they're all validated in `register_block_type_from_metadata`, so I'd rather keep it consistent.

Note that `register_block_type` actually calls `register_block_type_from_metadata` for block registration, so any block type registered via `register_block_type` will in fact undergo the validation in `register_block_type_from_metadata`: https://github.com/WordPress/wordpress-develop/blob/e005108fe18e6278f4e6dd0fec030e503fe0ff46/src/wp-includes/blocks.php#L584-L590

Finally, note that there's even an "inner" inconsistency with this PR, as validation of the allowed relative positions is still done in `register_block_type_from_metadata` 😅

Reply:
I left a https://github.com/WordPress/wordpress-develop/pull/5218#issuecomment-1720779818 with the details from my debugging after trying to use the test with the version before refactoring.

> Finally, note that there's even an "inner" inconsistency with this PR, as validation of the allowed relative positions is still done in register_block_type_from_metadata 😅

Yes, that could get improved, but we first need to settle on the responsibilities of processing `block.json` files. So far, it is mostly about mapping the JSON object to the matching fields in `WP_Block_Type`.

> AFAICS, no other field is being validated here in WP_Block_Type

There are some issues filed in Gutenberg to address it:
- https://github.com/WordPress/gutenberg/issues/48039
- https://github.com/WordPress/gutenberg/issues/48112